File: gcja2010QRA.hs
#!/usr/bin/env runghc
import Data.List.Split
main = getContents >>= mapM_ putStrLn . zipWith caseno [1..] . solve . lines
caseno i r = "Case #"++show i++": "++r
solve (c:cs) = map (toString . answer) . take (read c) . analyze $ cs ::[String]
analyze = map (\[c,_,p]->(read c, map read . words $ p)) . chunksOf 3
toString = unwords . map show . snd
answer (c,p) = maximum [(v,[i,j])| (i,a)<-ip, (j,b)<-ip, i<j, let v=a+b, v<=c]
  where ip = zip [1..] p

File: gcja2010QRA.py
#!/usr/bin/env python
import sys

def solve():
    C = int(sys.stdin.readline())
    I = int(sys.stdin.readline())
    P = map(int,sys.stdin.readline().split())
    c = [(v,[i+1,j+1]) for i in range(I) for j in range(i+1,I)
                       for v in [P[i]+P[j]] if v<=C]
    return " ".join(map(str,max(c)[1]))

for x in range(int(sys.stdin.readline())):
    print("Case #"+str(x+1)+": "+str(solve()))
